File: UIKit and Animations/WWDC 2013/Customizing Your App's Appearance for iOS 7/Sample_Code/TicTacToeApp/TicTacToe/TTTCountView.m
/*
     File: TTTCountView.m
 Abstract: 
 
  Version: 1.0
 
 Disclaimer: IMPORTANT:  This Apple software is supplied to you by Apple
 Inc. ("Apple") in consideration of your agreement to the following
 terms, and your use, installation, modification or redistribution of
 this Apple software constitutes acceptance of these terms.  If you do
 not agree with these terms, please do not use, install, modify or
 redistribute this Apple software.
 
 In consideration of your agreement to abide by the following terms, and
 subject to these terms, Apple grants you a personal, non-exclusive
 license, under Apple's copyrights in this original Apple software (the
 "Apple Software"), to use, reproduce, modify and redistribute the Apple
 Software, with or without modifications, in source and/or binary forms;
 provided that if you redistribute the Apple Software in its entirety and
 without modifications, you must retain this notice and the following
 text and disclaimers in all such redistributions of the Apple Software.
 Neither the name, trademarks, service marks or logos of Apple Inc. may
 be used to endorse or promote products derived from the Apple Software
 without specific prior written permission from Apple.  Except as
 expressly stated in this notice, no other rights or licenses, express or
 implied, are granted by Apple herein, including but not limited to any
 patent rights that may be infringed by your derivative works or by other
 works in which the Apple Software may be incorporated.
 
 The Apple Software is provided by Apple on an "AS IS" basis.  APPLE
 MAKES NO WARRANTIES, EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION
 THE IMPLIED WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY AND FITNESS
 FOR A PARTICULAR PURPOSE, REGARDING THE APPLE SOFTWARE OR ITS USE AND
 OPERATION ALONE OR IN COMBINATION WITH YOUR PRODUCTS.
 
 IN NO EVENT SHALL APPLE BE LIABLE FOR ANY SPECIAL, INDIRECT, INCIDENTAL
 OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION) ARISING IN ANY WAY OUT OF THE USE, REPRODUCTION,
 MODIFICATION AND/OR DISTRIBUTION OF THE APPLE SOFTWARE, HOWEVER CAUSED
 AND WHETHER UNDER THEORY OF CONTRACT, TORT (INCLUDING NEGLIGENCE),
 STRICT LIABILITY OR OTHERWISE, EVEN IF APPLE HAS BEEN ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.
 
 Copyright (C) 2013 Apple Inc. All Rights Reserved.
 
 
 Copyright © 2013 Apple Inc. All rights reserved.
 WWDC 2013 License
 
 NOTE: This Apple Software was supplied by Apple as part of a WWDC 2013
 Session. Please refer to the applicable WWDC 2013 Session for further
 information.
 
 IMPORTANT: This Apple software is supplied to you by Apple Inc.
 ("Apple") in consideration of your agreement to the following terms, and
 your use, installation, modification or redistribution of this Apple
 software constitutes acceptance of these terms. If you do not agree with
 these terms, please do not use, install, modify or redistribute this
 Apple software.
 
 In consideration of your agreement to abide by the following terms, and
 subject to these terms, Apple grants you a non-exclusive license, under
 Apple's copyrights in this original Apple software (the "Apple
 Software"), to use, reproduce, modify and redistribute the Apple
 Software, with or without modifications, in source and/or binary forms;
 provided that if you redistribute the Apple Software in its entirety and
 without modifications, you must retain this notice and the following
 text and disclaimers in all such redistributions of the Apple Software.
 Neither the name, trademarks, service marks or logos of Apple Inc. may
 be used to endorse or promote products derived from the Apple Software
 without specific prior written permission from Apple. Except as
 expressly stated in this notice, no other rights or licenses, express or
 implied, are granted by Apple herein, including but not limited to any
 patent rights that may be infringed by your derivative works or by other
 works in which the Apple Software may be incorporated.
 
 The Apple Software is provided by Apple on an "AS IS" basis. APPLE MAKES
 NO WARRANTIES, EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION THE
 IMPLIED WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY AND FITNESS FOR
 A PARTICULAR PURPOSE, REGARDING THE APPLE SOFTWARE OR ITS USE AND
 OPERATION ALONE OR IN COMBINATION WITH YOUR PRODUCTS.
 
 IN NO EVENT SHALL APPLE BE LIABLE FOR ANY SPECIAL, INDIRECT, INCIDENTAL
 OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION) ARISING IN ANY WAY OUT OF THE USE, REPRODUCTION,
 MODIFICATION AND/OR DISTRIBUTION OF THE APPLE SOFTWARE, HOWEVER CAUSED
 AND WHETHER UNDER THEORY OF CONTRACT, TORT (INCLUDING NEGLIGENCE),
 STRICT LIABILITY OR OTHERWISE, EVEN IF APPLE HAS BEEN ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.
 
 EA1002
 5/3/2013
 */

#import "TTTCountView.h"

@implementation TTTCountView

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.opaque = NO;
    }
    return self;
}

static CGFloat const TTTCountViewLineWidth = 1.0;
static CGFloat const TTTCountViewLineMargin = 4.0;
static NSInteger const TTTCountViewLineGroupCount = 5;

- (void)drawRect:(CGRect)rect
{
    [[self tintColor] set];
    CGRect bounds = self.bounds;
    CGFloat x = CGRectGetMaxX(bounds) - TTTCountViewLineWidth;
    for (NSInteger n = 0; n < self.count; n++) {
        x -= TTTCountViewLineMargin;
        if (((n + 1) % TTTCountViewLineGroupCount) == 0) {
            // Draw the diagonal line
            UIBezierPath *path = [UIBezierPath bezierPath];
            [path moveToPoint:CGPointMake(x + 0.5 * TTTCountViewLineWidth, CGRectGetMinY(bounds) + 0.5 * TTTCountViewLineWidth)];
            [path addLineToPoint:CGPointMake(x + 0.5 * TTTCountViewLineWidth + TTTCountViewLineGroupCount * TTTCountViewLineMargin, CGRectGetMaxY(bounds) - 0.5 * TTTCountViewLineWidth)];
            [path stroke];
        } else {
            // Draw a vertical line
            CGRect lineRect = bounds;
            lineRect.origin.x = x;
            lineRect.size.width = TTTCountViewLineWidth;
            UIRectFill(lineRect);
        }
    }
}

- (CGRect)rectForCount:(NSInteger)count
{
    CGRect bounds = self.bounds;
    CGRect rect = bounds;
    rect.size.width = TTTCountViewLineWidth + TTTCountViewLineMargin * count;
    rect.origin.x += bounds.size.width - rect.size.width;
    return rect;
}

- (void)setCount:(NSInteger)value
{
    if (_count != value) {
        CGRect oldRect = [self rectForCount:_count];
        _count = value;
        CGRect newRect = [self rectForCount:_count];
        CGRect dirtyRect = CGRectUnion(oldRect, newRect);
        [self setNeedsDisplayInRect:dirtyRect];
    }
}

- (void)tintColorDidChange
{
    [super tintColorDidChange];
    [self setNeedsDisplayInRect:[self rectForCount:self.count]];
}

#pragma mark - Accessibility

- (BOOL)isAccessibilityElement
{
    return YES;
}

- (NSString *)accessibilityLabel
{
    return [NSString stringWithFormat:@"%d", self.count];
}

- (UIAccessibilityTraits)accessibilityTraits
{
    return UIAccessibilityTraitImage;
}

@end

File: UIKit and Animations/WWDC 2013/Customizing Your App's Appearance for iOS 7/Sample_Code/TicTacToeApp_Swift/TicTacToeApp_Swift/View/TTTCountView.swift
//
//  TTTCountView.swift
//  TicTacToeApp_Swift
//
//  Created by Seyed Samad Gholamzadeh on 7/12/18.
//  Copyright © 2018 Seyed Samad Gholamzadeh. All rights reserved.
//

import UIKit

class TTTCountView: UIView {

	let TTTCountViewLineWidth: CGFloat = 1.0;
	let TTTCountViewLineMargin: CGFloat = 4.0;
	let TTTCountViewLineGroupCount: Int = 5;

	var count: Int! {
		didSet {
			var oldRect = CGRect.zero
			if oldValue != nil {
				oldRect = self.rect(for: oldValue)
			}
			let newRect = self.rect(for: self.count)
			let dirtyRect = oldRect.union(newRect)
			self.setNeedsDisplay(dirtyRect)
		}
	}
	

	override init(frame: CGRect) {
		super.init(frame: frame)
		 self.isOpaque = false
	}
	
	required init?(coder aDecoder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
	
	override func draw(_ rect: CGRect) {
		self.tintColor.set()
		let bounds = self.bounds
		var x = bounds.maxX - TTTCountViewLineWidth
		
		for n in 0..<self.count {
			x -= TTTCountViewLineMargin
			
			if CGFloat(n + 1).truncatingRemainder(dividingBy: CGFloat(TTTCountViewLineGroupCount)) == 0 {
				// Draw the diagonal line
				let path = UIBezierPath()
				path.move(to: CGPoint(x: x + 0.5 * TTTCountViewLineWidth, y: bounds.minY + 0.5 * TTTCountViewLineWidth))
				path.addLine(to: CGPoint(x: x + 0.5 * TTTCountViewLineWidth + CGFloat(TTTCountViewLineGroupCount) * TTTCountViewLineMargin, y: bounds.maxY - 0.5 * TTTCountViewLineWidth))
				path.stroke()
			}
			else {
				// Draw a vertical line
				var lineRect = bounds
				lineRect.origin.x = x
				lineRect.size.width = TTTCountViewLineWidth
				UIRectFill(lineRect)

			}
		}
	}
	
	func rect(for count: Int) -> CGRect {
		let bounds = self.bounds
		var rect = bounds
		rect.size.width = TTTCountViewLineWidth + TTTCountViewLineMargin * CGFloat(count)
		rect.origin.x += bounds.size.width - rect.size.width
		return rect
	}
	
	override func tintColorDidChange() {
		super.tintColorDidChange()
		self.setNeedsDisplay(self.rect(for: self.count))
	}
	
	//MARK: - Accessibility
	override var isAccessibilityElement: Bool {
		get {
			return true
		}
		
		set {
			super.isAccessibilityElement = newValue
		}
	}
	
	override var accessibilityLabel: String? {
		get {
			return "\(self.count)"
		}
		
		set {
			super.accessibilityLabel = newValue
		}
	}
	
	override var accessibilityTraits: UIAccessibilityTraits {
		get {
			return UIAccessibilityTraitImage
		}
		
		set {
			super.accessibilityTraits = newValue
		}
	}
	
	
}
